tests/src/rules/consistent-type-specifier-style.js

```diff
@@ -0,0 +1,291 @@
+import { RuleTester } from 'eslint';
+import { test, parsers, tsVersionSatisfies, eslintVersionSatisfies, typescriptEslintParserSatisfies } from '../utils';
+
+const rule = require('rules/consistent-type-specifier-style');
+
+const COMMON_TESTS = {
+  valid: [
+    //
+    // always valid
+    //
+    test({ code: "import Foo from 'Foo';" }),
+    test({ code: "import type Foo from 'Foo';" }),
+    test({ code: "import { Foo } from 'Foo';" }),
+    test({ code: "import { Foo as Bar } from 'Foo';" }),
+    test({ code: "import * as Foo from 'Foo';" }),
+    test({ code: "import 'Foo';" }),
+    test({ code: "import {} from 'Foo';" }),
+    test({ code: "import type {} from 'Foo';" }),
+
+    //
+    // prefer-top-level
+    //
+    {
+      code: "import type { Foo } from 'Foo';",
+      options: ['prefer-top-level'],
+    },
+    {
+      code: "import type { Foo as Bar } from 'Foo';",
+      options: ['prefer-top-level'],
+    },
+    {
+      code: "import type { Foo, Bar, Baz, Bam } from 'Foo';",
+      options: ['prefer-top-level'],
+    },
+  ],
+  invalid: [
+    //
+    // prefer-top-level
+    //
+    {
+      code: "import { type Foo } from 'Foo';",
+      output: "import type {Foo} from 'Foo';",
+      options: ['prefer-top-level'],
+      errors: [{
+        message: 'Prefer using a top-level type-only import instead of inline type specifiers.',
+        type: 'ImportDeclaration',
+      }],
+    },
+    {
+      code: "import { type Foo as Bar } from 'Foo';",
+      output: "import type {Foo as Bar} from 'Foo';",
+      options: ['prefer-top-level'],
+      errors: [{
+        message: 'Prefer using a top-level type-only import instead of inline type specifiers.',
+        type: 'ImportDeclaration',
+      }],
+    },
+    {
+      code: "import { type Foo, type Bar } from 'Foo';",
+      output: "import type {Foo, Bar} from 'Foo';",
+      options: ['prefer-top-level'],
+      errors: [{
+        message: 'Prefer using a top-level type-only import instead of inline type specifiers.',
+        type: 'ImportDeclaration',
+      }],
+    },
+    {
+      code: "import { Foo, type Bar } from 'Foo';",
+      output: "import { Foo,  } from 'Foo';\nimport type {Bar} from 'Foo';",
+      options: ['prefer-top-level'],
+      errors: [{
+        message: 'Prefer using a top-level type-only import instead of inline type specifiers.',
+        type: 'ImportSpecifier',
+      }],
+    },
+    {
+      code: "import { type Foo, Bar } from 'Foo';",
+      output: "import {  Bar } from 'Foo';\nimport type {Foo} from 'Foo';",
+      options: ['prefer-top-level'],
+      errors: [{
+        message: 'Prefer using a top-level type-only import instead of inline type specifiers.',
+        type: 'ImportSpecifier',
+      }],
```

Comment:
ohh that's a really good test case I forgot about.
this one is even better though:
```ts
import AValue, { type Foo } from 'Foo';
```
Because it should fix to
```ts
import AValue from 'Foo';
import type Foo from 'Foo';
```
i.e. we have to remove the empty `{}` as well

I wonder... @ljharb - do you have a preference here?
Should there be a new rule which enforces against empty named specifier blocks, or should we just bring that complexity into each lint rule?

Reply:
@bradzacher Isn't that invalid syntax? `import {} from` is already invalid.

If it's valid to have `import type {} from` then i think we just need to handle that in individual rules.

(Sorry for the delayed response; i'd like to land this soon so I'll keep a closer eye on it)

Reply:
> Isn't that invalid syntax? import {} from is already invalid

According to [the spec](https://262.ecma-international.org/6.0/#sec-imports) `{}` is allowed:

```
NamedImports :
    { }
    { ImportsList }
    { ImportsList , }
```

Based on testing - all eslint parsers allow it.
Also the flow/hermes/typescript/babel parsers allow `import type {}`

Reply:
oh oof, gross. ok then i think a) we'd need a separate rule to ban empty named import/export blocks, and b) this rule's autofixer should just always be removing empty blocks when removing duplicates.

Reply:
there is *some* additional complexity that will get added to the fixer for this, but we can definitely do it!

The other option is to let this fixer purely focus on moving specifiers around, and then let the new `no-empty-named-blocks` focus on cleaning up the empty curlies - similar to how the fixer right now does not attempt to consolidate imports, and just moves things around.

Reply:
I suppose that's an option too - but I think in the common case people will not want empty braces, and it's nice when autofixers optimize for the common case.

That said, certainly if someone _wants_ empty braces for some reason, that's only possible if this autofixer leaves them behind :-/